### PR TITLE
Update query builder data selector for datasets

### PR DESCRIPTION
**Pull request by kulyk**
This PR adds datasets support for GUI query data source picker (feature epic #18659)

before-after-img

When people have at least one dataset in Metabase, we're going to display a new first step. Otherwise, the picker should remain the same. About the "tabs":

* **Saved Questions** works the same way it does now
* **Raw Data** contains databases, schemas, and tables (the one you can see when using the picker now). The only difference here is that Raw Data doesn't list the "Saved Questions" option as they have their own tab
* **Datasets** tab looks the same way as the saved question picker, but only lists datasets

## Implementation Details

The goal is to have a minimal version of datasets on `master` ASAP, so we can play around with the whole experience and introduce changes early. There are some trade-offs I had to address, and the main principles in the decisions are speed and YAGNI in case we want to change something about the requirements.

### Reusing saved questions picker

Datasets picker is expected to look and work the same way as the saved question picker, but only display datasets. So we should either extend the existing saved questions picker to support datasets too (datasets are saved questions under the hood, so their APIs are almost identical) or extract some common code and create two different components. This PR _extends_ the saved questions picker, so the data selector uses _just one_ component for both datasets and saved questions.

### Encoding extra parameters in schema ID

As mentioned above, this PR actually uses the saved questions picker for datasets too. When a user selects a collection from the picker, [`SavedQuestionList` requests a list of questions from the backend using the Schema entity loader](https://github.com/metabase/metabase/blob/5fcf4a3aa216298ec12a2d380668df0ab33d9e16/frontend/src/metabase/query_builder/components/saved-question-picker/SavedQuestionList.jsx#L50).

The trick here is that the BE will only return saved questions (not datasets) as expected. And we need to use another API endpoint to get datasets from a particular collection. So we need a way to let the Schema entity loader know when we need questions and when we need datasets. 

This PR extends the schema ID string to also contain some extra JSON data. Schema IDs currently look like `14:Application` where `14` is the database's ID and `Application` is a name of a schema. For saved questions, schema IDs look like `-1337:Marketing` where `-1337` is a "virtual" database with saved questions and `Marketing` is a name of a collection. This PR extends the format, so we can now also have something like `-1337:Marketing:{"isDatasets":true}`. Schema entity later [uses this parameter](https://github.com/metabase/metabase/blob/139ee2cea6460cc0a4183ddc59d86de9922d37b2/frontend/src/metabase/entities/schemas.js#L41) to decide which endpoint it should use.

### Schema file reorg

There is a [`schema` file in FE source code root](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/schema.js). It has entity schema definitions and some extra helpers for encoding and decoding schema ID, etc. I needed to use `generateSchemaId` in `metabase/lib/saved-questions` ([here](https://github.com/metabase/metabase/blob/139ee2cea6460cc0a4183ddc59d86de9922d37b2/frontend/src/metabase/lib/saved-questions/saved-questions.js#L19)), but the FE broke because some intertwined imports order if I just imported it. So in this PR, the schema utils are moved to [`metabase/lib/schema`](https://github.com/metabase/metabase/blob/139ee2cea6460cc0a4183ddc59d86de9922d37b2/frontend/src/metabase/lib/schema/schema.js). Entity schemas are still kept in `metabase/schema`

## To Verify

1. Archive any datasets if you have any
2. Ask a question > Simple / Custom Question
3. Make sure the data source picker looks and works the same way it does on `master`
4. Turn any question into a dataset (you can follow the steps from #18702)
5. Ask a question > Simple / Custom Question
6. The picker should have a new look now. You should see three options: Datasets, Raw Data, Saved Questions. Don't click anything for now

**Common Search**

Try to find a dataset, a table, and a saved question by their names. The search shouldn't skip any of these models

**Raw Data Tab**

1. Click the "Raw Data" tab
2. You should not see "Saved Questions" DB anywhere in the list
3. Try to find a dataset, a table, and a saved question by their names. While being on the "Raw Data" tab, the search should be scoped to tables. So you shouldn't see any datasets and saved questions in the results
4. Make sure the picker works the same way it does on `master` (expanding DB schemas, selecting things, etc.)
5. Click the back button, you should appear on the list with "Datasets", "Raw Data" and "Saved Questions"

**Datasets Tab**

1. Click the "Datasets" tab
2. Try to find a dataset, a table, and a saved question by their names. While being on the "Datasets" tab, the search should be scoped to datasets. So you shouldn't see any tables and saved questions in the results
3. Click on a few collections. If a collection has datasets, they should be shown on the right side
4. Make sure you don't see saved questions in the collections

**Saved Questions Tab**

1. Click the "Saved Questions" tab
2. Try to find a dataset, a table, and a saved question by their names. While being on the "Saved Questions" tab, the search should be scoped to saved questions only. So you shouldn't see any datasets and tables in the results
3. Make sure the picker works the same way it does on `master` (clicking collections, displaying content, etc.)
4. Make sure you don't see a dataset in the collection you saved it to
5. Click the back button, you should appear on the list with "Datasets", "Raw Data" and "Saved Questions"

**Ask a question!**

Create and save a question based on a dataset

## Demo

https://user-images.githubusercontent.com/17258145/139733530-ea3182f6-a820-407d-b8db-3490859fc061.mov

### CODE REVIEW

**Comment by mazameli** [comment]
Clicking on New Question > Simple, then New Question > Custom, then New Question > Simple again seems to prevent the data selector from appearing: https://share.getcloudapp.com/12uPKOyP

Here's the stacktrack I'm seeing:

```
Step null not found in BUCKET,DATABASE,SCHEMA,TABLE. 
    at UnconnectedDataSelector (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:287117:5)
    at DataSelector (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:287095:5)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at EntityListLoader (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:255858:5)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at _class2 (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:258292:9)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:256669:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at Connect(DataSelector)
    at DatabaseSchemaAndTableDataSelector
    at div
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at div
    at Tag (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:42826:5)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at NotebookCellItem (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:309712:23)
    at div
    at Tag (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:42826:5)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at DataStep (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:311392:20)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at div
    at Tag (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:42826:5)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at div
    at Tag (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:42826:5)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at div
    at Tag (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:42826:5)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at div
    at ExpandingContent (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:215874:32)
    at NotebookStep (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:310157:5)
    at div
    at NotebookSteps (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:310851:5)
    at div
    at Tag (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:42826:5)
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at Notebook (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:309494:24)
    at div
    at StyledComponent (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300749:9)
    at QueryViewNotebook (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:319684:38)
    at div
    at div
    at div
    at View (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:319302:5)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:216287:9)
    at QueryBuilder (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:323632:5)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:257397:7)
    at http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259313:27
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259095:9)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259095:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:339450:23
    at UserAuthWrapper (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:272703:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at UserAuthWrapper (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:272703:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at div
    at ScrollToTop (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:105162:5)
    at <anonymous> (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:60900:37)
    at App (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:163122:5)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259095:9)
    at <anonymous> (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:60900:37)
    at <anonymous> (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:60900:37)
    at ThemeProvider (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300434:5)
    at DragDropContextProviderImpl (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:204900:28)
    at Provider (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247541:26)
overrideMethod @ react_devtools_backend.js:2528
getPreviousStep @ DataSelector.jsx:504
renderActiveStep @ DataSelector.jsx:785
render @ DataSelector.jsx:942
finishClassComponent @ react-dom.development.js:17160
updateClassComponent @ react-dom.development.js:17110
beginWork @ react-dom.development.js:18620
beginWork$1 @ react-dom.development.js:23179
performUnitOfWork @ react-dom.development.js:22154
workLoopSync @ react-dom.development.js:22130
performSyncWorkOnRoot @ react-dom.development.js:21756
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
discreteUpdates$1 @ react-dom.development.js:21893
discreteUpdates @ react-dom.development.js:806
dispatchDiscreteEvent @ react-dom.development.js:4168
react-dom.development.js:2962 Uncaught Error: Unable to find node on an unmounted component.
    at findCurrentFiberUsingSlowPath (react-dom.development.js:2962)
    at findCurrentHostFiber (react-dom.development.js:3124)
    at findHostInstanceWithWarning (react-dom.development.js:24293)
    at Object.findDOMNode (react-dom.development.js:24804)
    at Popover._getTargetElement (Popover.jsx:287)
    at Popover.jsx:311
    at callback (react-dom.development.js:24752)
    at callCallback (react-dom.development.js:12490)
    at commitUpdateQueue (react-dom.development.js:12511)
    at commitLifeCycles (react-dom.development.js:19883)
findCurrentFiberUsingSlowPath @ react-dom.development.js:2962
findCurrentHostFiber @ react-dom.development.js:3124
findHostInstanceWithWarning @ react-dom.development.js:24293
findDOMNode @ react-dom.development.js:24804
_getTargetElement @ Popover.jsx:287
(anonymous) @ Popover.jsx:311
callback @ react-dom.development.js:24752
callCallback @ react-dom.development.js:12490
commitUpdateQueue @ react-dom.development.js:12511
commitLifeCycles @ react-dom.development.js:19883
commitLayoutEffects @ react-dom.development.js:22803
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
commitRootImpl @ react-dom.development.js:22541
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
commitRoot @ react-dom.development.js:22381
finishSyncRender @ react-dom.development.js:21807
performSyncWorkOnRoot @ react-dom.development.js:21793
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
discreteUpdates$1 @ react-dom.development.js:21893
discreteUpdates @ react-dom.development.js:806
dispatchDiscreteEvent @ react-dom.development.js:4168
react_devtools_backend.js:2528 The above error occurred in one of your React components:

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
overrideMethod @ react_devtools_backend.js:2528
logCapturedError @ react-dom.development.js:19527
logError @ react-dom.development.js:19564
update.callback @ react-dom.development.js:20708
callCallback @ react-dom.development.js:12490
commitUpdateQueue @ react-dom.development.js:12511
commitLifeCycles @ react-dom.development.js:19883
commitLayoutEffects @ react-dom.development.js:22803
callCallback @ react-dom.development.js:188
invokeGuardedCallbackDev @ react-dom.development.js:237
invokeGuardedCallback @ react-dom.development.js:292
commitRootImpl @ react-dom.development.js:22541
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
commitRoot @ react-dom.development.js:22381
finishSyncRender @ react-dom.development.js:21807
performSyncWorkOnRoot @ react-dom.development.js:21793
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
discreteUpdates$1 @ react-dom.development.js:21893
discreteUpdates @ react-dom.development.js:806
dispatchDiscreteEvent @ react-dom.development.js:4168
react-dom.development.js:248 Uncaught Error: A cross-origin error was thrown. React doesn't have access to the actual error object in development. See https://fb.me/react-crossorigin-error for more information.
    at Object.invokeGuardedCallbackDev (react-dom.development.js:248)
    at invokeGuardedCallback (react-dom.development.js:292)
    at commitRootImpl (react-dom.development.js:22541)
    at unstable_runWithPriority (scheduler.development.js:653)
    at runWithPriority$1 (react-dom.development.js:11039)
    at commitRoot (react-dom.development.js:22381)
    at finishSyncRender (react-dom.development.js:21807)
    at performSyncWorkOnRoot (react-dom.development.js:21793)
    at react-dom.development.js:11089
    at unstable_runWithPriority (scheduler.development.js:653)
invokeGuardedCallbackDev @ react-dom.development.js:248
invokeGuardedCallback @ react-dom.development.js:292
commitRootImpl @ react-dom.development.js:22541
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
commitRoot @ react-dom.development.js:22381
finishSyncRender @ react-dom.development.js:21807
performSyncWorkOnRoot @ react-dom.development.js:21793
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
discreteUpdates$1 @ react-dom.development.js:21893
discreteUpdates @ react-dom.development.js:806
dispatchDiscreteEvent @ react-dom.development.js:4168
react_devtools_backend.js:2528 Step null not found in BUCKET,DATABASE,SCHEMA,TABLE. 
    at UnconnectedDataSelector (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:287117:5)
    at DataSelector (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:287095:5)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at EntityListLoader (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:255858:5)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at _class2 (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:258292:9)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:256669:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at Connect(DataSelector)
    at DatabaseSchemaAndTableDataSelector
    at QuestionDataSelector (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:317597:20)
    at div
    at div
    at View (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:319302:5)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:216287:9)
    at QueryBuilder (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:323632:5)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:257397:7)
    at http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259313:27
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259095:9)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259095:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:339450:23
    at UserAuthWrapper (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:272703:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at UserAuthWrapper (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:272703:9)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at div
    at ScrollToTop (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:105162:5)
    at <anonymous> (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:60900:37)
    at App (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:163122:5)
    at Connect (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247707:28)
    at _class (http://localhost:8080/app/dist/app-main.hot.bundle.js?5beb603ec1208c29be63:259095:9)
    at <anonymous> (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:60900:37)
    at <anonymous> (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:60900:37)
    at ThemeProvider (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:300434:5)
    at DragDropContextProviderImpl (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:204900:28)
    at Provider (http://localhost:8080/app/dist/vendor.hot.bundle.js?063039715087fe38941f:247541:26)
overrideMethod @ react_devtools_backend.js:2528
getPreviousStep @ DataSelector.jsx:504
renderActiveStep @ DataSelector.jsx:785
render @ DataSelector.jsx:942
finishClassComponent @ react-dom.development.js:17160
updateClassComponent @ react-dom.development.js:17110
beginWork @ react-dom.development.js:18620
beginWork$1 @ react-dom.development.js:23179
performUnitOfWork @ react-dom.development.js:22154
workLoopSync @ react-dom.development.js:22130
performSyncWorkOnRoot @ react-dom.development.js:21756
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
scheduleUpdateOnFiber @ react-dom.development.js:21199
enqueueSetState @ react-dom.development.js:12639
Component.setState @ react.development.js:471
onStateChange @ connectAdvanced.js:207
notify @ Subscription.js:23
notifyNestedSubs @ Subscription.js:62
onStateChange @ connectAdvanced.js:204
dispatch @ createStore.js:173
(anonymous) @ middleware.js:22
(anonymous) @ index.js:28
(anonymous) @ store.js:25
dispatch @ applyMiddleware.js:35
action @ store.js:20
_callee3$ @ actions.js:535
tryCatch @ runtime.js:63
invoke @ runtime.js:293
(anonymous) @ runtime.js:118
asyncGeneratorStep @ selectors.js:36
_next @ selectors.js:36
Promise.then (async)
asyncGeneratorStep @ selectors.js:36
_next @ selectors.js:36
(anonymous) @ selectors.js:36
(anonymous) @ selectors.js:36
(anonymous) @ actions.js:321
(anonymous) @ store.js:23
(anonymous) @ bindActionCreators.js:3
UNSAFE_componentWillMount @ QueryBuilder.jsx:177
callComponentWillMount @ react-dom.development.js:12953
mountClassInstance @ react-dom.development.js:13048
updateClassComponent @ react-dom.development.js:17101
beginWork @ react-dom.development.js:18620
beginWork$1 @ react-dom.development.js:23179
performUnitOfWork @ react-dom.development.js:22154
workLoopSync @ react-dom.development.js:22130
performSyncWorkOnRoot @ react-dom.development.js:21756
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
discreteUpdates$1 @ react-dom.development.js:21893
discreteUpdates @ react-dom.development.js:806
dispatchDiscreteEvent @ react-dom.development.js:4168
Show 5 more frames
react-dom.development.js:2962 Uncaught Error: Unable to find node on an unmounted component.
    at findCurrentFiberUsingSlowPath (react-dom.development.js:2962)
    at findCurrentHostFiber (react-dom.development.js:3124)
    at findHostInstanceWithWarning (react-dom.development.js:24293)
    at Object.findDOMNode (react-dom.development.js:24804)
    at _class._getElement (ExplicitSize.jsx:31)
    at ExplicitSize.jsx:101
    at throttled (throttle.js:32)
```

**Comment by mazameli** [comment]
Minor polish: the datasets icon in the picker looks small; 16px is desired.

image

**Comment by mazameli** [comment]
If I use the datapicker's search to find a dataset, when I select it, it selects some other table altogether: https://share.getcloudapp.com/kpu6NyDk

Or it doesn't let me select it at all — in this video I have a dataset called "a guy" that I'm trying to click on here: https://share.getcloudapp.com/BluDQpxd

**Comment by mazameli** [comment]
When clicking on the "Saved questions" option _the first time_, it incorrectly takes me to the raw data list of databases. Going back in the picker and then selecting "Saved questions" again works correctly the next time.

https://share.getcloudapp.com/6quvx0vd

**Comment by mazameli** [comment]
We should use the datasets/waffle icon in this context, too: 
image

**Comment by kulyk** [comment]
There is an issue with the picker behavior for questions built on datasets. Steps to reproduce:

1. Create and save a question based on a dataset
2. Open query notebook editor
3. Click the dataset name in the "Data" step at the top
4. The "Saved questions" picker open without anything marked as selected

**Expected:** datasets picker should be open with selected dataset's collection open and the dataset itself has to be highlighted as selected

This happens because datasets have the same "virtual" table ID as save questions (`card__{{ question_id }}`), so the picker can't figure out which picker to show. I'd better address this in a different PR not to overload this one

**Review by mazameli — APPROVED** [review]
You fixed all the issues I'd previously found!

File: frontend/test/metabase/scenarios/question/datasets.cy.spec.js [hunk 1]
```diff
@@ -26,6 +26,99 @@ describe("scenarios > datasets", () => {
...
+          tables: true,
+        });
+
+        cy.findByText("Datasets").click();
+        cy.findByTestId("select-list").within(() => {
+          cy.findByText("Orders");
+          cy.findByText("Orders, Count").should("not.exist");
+        });
+        testDataPickerSearch({
+          inputPlaceholderText: "Search for a model...",
```
Comment (mazameli) [inline frontend/test/metabase/scenarios/question/datasets.cy.spec.js:55]:
```suggestion
          inputPlaceholderText: "Search for a dataset…",
```

File: frontend/test/metabase/scenarios/question/datasets.cy.spec.js [hunk 2]
```diff
@@ -26,6 +26,99 @@ describe("scenarios > datasets", () => {
...
+        });
+        cy.icon("chevronleft").click();
+
+        cy.findByText("Saved Questions").click();
+        cy.findByTestId("select-list").within(() => {
+          cy.findByText("Orders, Count");
+          cy.findByText("Orders").should("not.exist");
+        });
+        testDataPickerSearch({
+          inputPlaceholderText: "Search for a question...",
```
Comment (mazameli) [inline frontend/test/metabase/scenarios/question/datasets.cy.spec.js:67]:
```suggestion
          inputPlaceholderText: "Search for a question…",
```

File: frontend/test/metabase/scenarios/question/datasets.cy.spec.js [hunk 3]
```diff
@@ -26,6 +26,99 @@ describe("scenarios > datasets", () => {
...
+          query: "Ord",
+          cards: true,
+        });
+        cy.icon("chevronleft").click();
+
+        cy.findByText("Raw Data").click();
+        cy.findByText("Sample Dataset");
+        cy.findByText("Saved Questions").should("not.exist");
+        testDataPickerSearch({
+          inputPlaceholderText: "Search for a table...",
```
Comment (mazameli) [inline frontend/test/metabase/scenarios/question/datasets.cy.spec.js:77]:
```suggestion
          inputPlaceholderText: "Search for a table…",
```

File: frontend/test/metabase/scenarios/question/new.cy.spec.js [hunk 4]
```diff
@@ -187,7 +187,7 @@ describe("scenarios > question > new", () => {
       });
 
       it("should perform a search scoped to saved questions", () => {
-        cy.findByPlaceholderText("Search for a question").type("Grouped");
+        cy.findByPlaceholderText("Search for a question...").type("Grouped");
```
Comment (mazameli) [inline frontend/test/metabase/scenarios/question/new.cy.spec.js:190]:
```suggestion
        cy.findByPlaceholderText("Search for a question…").type("Grouped");
```

File: frontend/test/metabase/scenarios/question/new.cy.spec.js [hunk 5]
```diff
@@ -230,7 +230,7 @@ describe("scenarios > question > new", () => {
...
```
Comment (mazameli) [inline frontend/test/metabase/scenarios/question/new.cy.spec.js:233]:
```suggestion
        cy.findByPlaceholderText("Search for a question…").type("Grouped");
```

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep.unit.spec.js [hunk 6]
```diff
@@ -154,7 +154,7 @@ describe("Notebook Editor > Join Step", () => {
   }
 
   async function selectTable(tableName) {
-    fireEvent.click(screen.queryByText(/Sample Dataset/i));
+    await waitForElementToBeRemoved(() => screen.queryByText("Loading..."));
```
Comment (mazameli) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep.unit.spec.js:157]:
```suggestion
    await waitForElementToBeRemoved(() => screen.queryByText("Loading…"));
```

File: frontend/test/metabase/scenarios/question/datasets.cy.spec.js [hunk 7]
```diff
@@ -26,6 +26,99 @@ describe("scenarios > datasets", () => {
...
+      cy.request("PUT", "/api/card/1", { dataset: true });
+    });
+
+    it("transforms the data picker", () => {
+      cy.visit("/question/new");
+      cy.findByText("Custom question").click();
+
+      popover().within(() => {
+        testDataPickerSearch({
+          inputPlaceholderText: "Search for some data...",
```
Comment (mazameli) [inline frontend/test/metabase/scenarios/question/datasets.cy.spec.js:42]:
```suggestion
          inputPlaceholderText: "Search for some data…",
```

**Comment by kulyk** [comment]
Thanks for the ellipsis fix @mazameli, I batched them in a single commit 🙏

**Review by flamber — COMMENTED** [review]
1. Should it be possible to change a Datasets into a Question again, or is it just one-way, like converting QB to SQL?
2. Minor double menu issue, when using join:

https://user-images.githubusercontent.com/1447303/139862909-b22b2077-befa-4dc8-b5e3-115c2810d23d.mp4

**Comment by kulyk** [comment]
> 1. Should it be possible to change a Datasets into a Question again, or is it just one-way, like converting QB to SQL?
> 2. Minor double menu issue, when using join

1. Yes, but it's not implemented yet
2. Not sure I can see what's wrong there, can you explain please?

**Comment by flamber** [comment]
@kulyk Is it supposed to show "< Raw Data", when I clicked back (time 00.07), instead of just showing all three options (time 00.19)?

**Comment by kulyk** [comment]
> @kulyk Is it supposed to show "< Raw Data", when I clicked back (time 00.07), instead of just showing all three options (time 00.19)?

Oh, you're right, I think we should just skip to the first view 👍

File: frontend/src/metabase/query_builder/components/notebook/steps/JoinStep.unit.spec.js [hunk 8]
```diff
@@ -414,6 +423,7 @@ describe("Notebook Editor > Join Step", () => {
       expect(screen.queryAllByText("Pick a column...")).toHaveLength(2);
     });
 
+    // ###
```
Comment (daltojohnso) [inline frontend/src/metabase/query_builder/components/notebook/steps/JoinStep.unit.spec.js:426]: random comment

**Review by daltojohnso — COMMENTED** [review]
Changes look solid. `DataSelector.jsx` makes me weep. Two things to note:

1. The "Raw Data" back button has a lot of dead space where clicking does nothing.

https://user-images.githubusercontent.com/13057258/139942365-d4154264-f379-4284-bf00-f1ffb15255b8.mov

2. Looks like the request to fetch "Your personal collection" 404s now. The GET is sent to `/api/database/-1337/datasets/Your%20personal%20collection` and I think (not 100% sure) that it needs to go to a URL like `/api/database/-1337/datasets/First%20Last's%20Personal%20Collection`

Overall, however, your code looks good, so I won't explicitly block on these bugs, so feel free to merge after you fix.

**Comment by kulyk** [comment]
Great catches @daltojohnso, thank you 🙏  Both issues addressed, merging

**Pull request merged**